crypto/vrf_go.go

```diff
@@ -0,0 +1,465 @@
+	// inlined vrf_expand_sk
+	Y, xScalar, truncatedHashedSkString, err := sk.expand()
+	if err != nil {
+		// TODO: this method should return an error.
+		// fmt.Println("issue expanding:", err)
+		return proof, false
+	}
+	proof, err = pureGoVrfProve(Y, xScalar, truncatedHashedSkString, msg)
+	if err != nil {
+		// TODO: this method should return an error.
+		// fmt.Println("issue proving:", err)
+		return proof, false
+	}
+	return proof, err == nil
+}
+
+/* Construct a proof for a message alpha per draft spec section 5.1.
+ * Takes in a secret scalar x, a public point Y, and a secret string
+ * truncated_hashed_sk that is used in nonce generation.
+ * These are computed from the secret key using the expand_sk function.
+ * Constant time in everything except the length of alpha.
+ */
+func pureGoVrfProve(Y *edwards25519.Point, xScalar *edwards25519.Scalar, truncHashedSk []byte, alpha []byte) (VrfProof, error) {
+	var pi VrfProof
+	H, err := vrfHashToCurveElligator225519(Y, alpha)
+	if err != nil {
+		return VrfProof{}, err
+	}
+	Gamma := edwards25519.NewIdentityPoint()
+	Gamma.ScalarMult(xScalar, H)
+
+	kScalar := vrfNonceGeneration(truncHashedSk, H)
+	kB := edwards25519.NewIdentityPoint()
+	kB.ScalarBaseMult(kScalar)
+	kH := edwards25519.NewIdentityPoint()
+	kH.ScalarMult(kScalar, H)
+
+	cScalar := vrfHashPoints(H, Gamma, kB, kH)
+	s := edwards25519.NewScalar()
+	s.MultiplyAdd(cScalar, xScalar, kScalar)
+
+	// output pi
+	copy(pi[:], Gamma.Bytes())
+	copy(pi[32:], cScalar.Bytes()[:16])
+	copy(pi[48:], s.Bytes())
+	return pi, nil
+}
+
+/* Hash a message to a curve point using Elligator2.
+ * Specified in VRF draft spec section 5.4.1.2.
+ * The actual elligator2 implementation is ge25519_from_uniform.
+ * Runtime depends only on alphalen (the message length)
+ */
+func vrfHashToCurveElligator225519(Y *edwards25519.Point, alpha []byte) (*edwards25519.Point, error) {
+	hs := sha512.New()
+
+	hs.Write([]byte{vrfSuite})
+	hs.Write([]byte{1})
+	hs.Write(Y.Bytes())
+	hs.Write(alpha)
+	rString := hs.Sum(nil)
+	rString[31] &= 0x7f // clear sign bit
+
+	hBytes, err := ge25519FromUniform(rString)
+	if err != nil {
+		return nil, err
+	}
+	result := &edwards25519.Point{}
+	result.SetBytes(hBytes[:]) // ge25519_frombytes(&H_point, h_string);
+	return result, nil
+}
+
+// elligator2
+func ge25519FromUniform(r []byte) ([]byte, error) {
+	s := make([]byte, 32)
+	var e, negx, rr2, x, x2, x3 *field.Element
+	var p3 *edwards25519.Point
+	var eIsMinus1 int
+	var xSign byte
+
+	copy(s, r)
+	xSign = s[31] & 0x80
+	s[31] &= 0x7f
+
+	rr2 = &field.Element{}
+	rr2.SetBytes(s) // fe25519_frombytes(rr2, s);
+
+	// elligator
+	rr2.Square(rr2) // fe25519_sq2(rr2, rr2);
+	rr2.Add(rr2, rr2)
+	rr2Bytes := rr2.Bytes()
+	rr2Bytes[0]++
+	rr2.SetBytes(rr2Bytes) // rr2[0]++;
+	rr2.Invert(rr2)        // fe25519_invert(rr2, rr2);
+
+	x = &field.Element{}
+
+	const curve25519A = 486662
+	curve25519AElement := (&field.Element{}).One().Mult32((&field.Element{}).One(), curve25519A)
```

Comment:
```suggestion
	one := new(field.Element).One()
	curve25519AElement := new(field.Element).Mult32(one, curve25519A)
```

Also, this could be a global, so you only compute it once.